text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+
+We need to be able to represent the results of processing a stream of unprocessed data as a stream of data frames.
+
+To be able to output a data frame as a stream, we need to know two key elements: that the data frame is incomplete as-is, and a unique identifier that allows us to later stream additional data for this frame.
+
+To accomplish this streaming, we also introduce an `EndFrame`. Frame and EndFrame work together to allow a frame to be streamed as a multi-part frame, ending once the corresponding EndFrame has been read.
+
+As an example, let's say we were processing some content and wanted to output the first row and later the second row of this table:
+
+| tag  | length |
+| ---- | ------ |
+| head | 1024   |
+| body | 8192   |
+
+To do this, we would send the two separate frames, both marked as partial frames.
+
+```rust
+let frame_id = Uuid::new_v4();
+output.send(UntaggedValue::DataFrame {
+    headers: Some(vec!["tag", "length"]),
+    length: vec![
+        vec![ Value::from("head"), Value::from(1024)]
+    ],
+    partial_frame_id: Some(frame_id),
+    is_object_false,
+}.into_value());
+
+// ... time passes
+
+output.send(UntaggedValue::DataFrame {
+    headers: Some(vec!["tag", "length"]),
+    length: vec![
+        vec![ Value::from("body"), Value::from(8192)]
+    ],
+    partial_frame_id: Some(frame_id),
+    is_object_false,
+}.into_value());
+
+output.send(UntaggedValue::EndFrame(frame_id).into_value());
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Much of the implementation is part of the explanation in the previous section. In this section, we'll explore more of the impact of making this change.
+
+`UntaggedValue` will have `Table` and `Row` replaced with `DataFrame` (possibly just called `Frame` for brevity) and `EndFrame`.
+
+All commands that operate on Row today will need to be updated to work with the data frame instead.
+
+Rather than operating on a single row, commands will need to be updated to handle a frame at a time. Here, the mapping should largely be the same, though an additional inner loop to process over the rows will be necessary. This processing can be done serial or in parallel and may be done synchronously or asynchronously.
+
+Commands that worked over inner tables should be able to migrate to data frames, as there is a strong overlap in functionality.
+
+Commands that filter will work similarly as before, and may opt to output streams which are flattened by the output stream. This allows them to optionally return no frame if no rows in the frame met the requirements of the filter.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Some drawbacks come to mind:
+
+- This is a large, non-trivial amount of work. Getting this landed, updating the commands to use the new model, and thoroughly testing will take time.
```

Comment:
Totally agree, this is like replacing the engine of our car. We would definitely need a good plan on how to impelment this, perhaps in bite-sized chunks if possible. 

Would it be possible to perhaps implement a dataframe plugin to smooth out the rough edges before diving into the deep end? 

Maybe even, this plugin could somehow work as a shim between worlds translating current state to dataframe state and therefore being able to be used stop breaking changes from off-world plugins - kind of a translation layer.

Reply:
I don't see a way to bolt this on, unfortunately. I mentioned to @thegedge in another comment we can have Row and Frame live together for a while, and then phase out Row as we transition. But enough of the other systems will need to know about Frame that it's difficult to add this outside of core set of concepts.